### PR TITLE
Update CircleCI, readthedocs, and Docker configuration

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

This pull request refreshes our readthedocs and CircleCI configuration.

### readthedocs
Switching to mamba for building the documentation on readthedocs:
- simplifies dependency management, because the documentation does no longer need a separate list of dependencies
- enables setting the version number using git (see #1578)
- slows down the documentation build

This pull request also moves the test for warnings when building the documentation for branches from CircleCI to readthedocs, reducing some test duplication.

### CircleCI
- Test and installation from source jobs are now twice as fast
- Use Mamba instead of Conda for installation
- Correctly store test results, so we enable the analysis features provided by CircleCI
- Compress artifacts, because uploading the test results took almost as long as running them
- Use appropriately sized machines for each job
- Unpin flake8
- Remove duplicated code from the configuration file

### Docker
- Use Mamba instead of Conda for installation

Link to documentation: https://esmvaltool--1588.org.readthedocs.build/projects/ESMValCore/en/1588/contributing.html#documentation

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: .readthedocs.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# .readthedocs.yaml
+# Read the Docs configuration file
+# See https://docs.readthedocs.io/en/stable/config-file/v2.html for details
+
+# Required
+version: 2
+
+# Set the version of Python and other tools you might need
+build:
+  os: ubuntu-22.04
```
Comment (valeriupredoi) [inline .readthedocs.yaml:10]: I think 14.04 is also fine :rofl:

File: setup.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
     # Use with pip install . to install from source
     'install': [
         'cartopy',
-        'cf-units>=3.0.0',
+        'cf-units>=3.0.0,!=3.0.1.post0',
```
Comment (valeriupredoi) [inline setup.py:34]: can you add a short comment here why the blocked post0 pls - so we know when/why to remove it in the future

**Review by valeriupredoi — APPROVED** [review]
this is a very beautiful PR @bouweandela - awesome (am particularly happy about the CI enhancement!), why is it still draft?

File: setup.py (hunk 2 above)
Comment (bouweandela) [reply setup.py:35]: done

**Comment by bouweandela** [comment]
Ready for review now @valeriupredoi! Is this title any better?

**Comment by valeriupredoi** [comment]
beautiful title, beautiful PR :beer: Merge at will @bouweandela - only two question from me - are we OK with unpinning `flake8` now? and do we want to fix `Docker.exp` in Tool next - now would be a good time since you've already polished the docker file here

**Comment by bouweandela** [comment]
> are we OK with unpinning flake8 now?

The tests pass just fine, so I don't see a problem with unpinning it.

> do we want to fix Docker.exp in Tool next - now would be a good time since you've already polished the docker file here

It might be nice to have a look at reducing the size of the containers at some point, e.g. we could stop including our entire GitHub repo in the Docker container (saves about 150 MB), but I wasn't able to do that here because it's used by that container and I didn't want to break it. Note that that container is pretty much broken by design, because it uses an ESMValCore version that is not compatible with the ESMValTool version: the ESMValTool development version is compatible with the latest released version of ESMValCore.

**Comment by valeriupredoi** [comment]
no - the flak8 was pinned not for test fail resons but for all that garbage that'd print to screen in case some other test was failing, remember we were trying to figure out why we'd not get the correct sydout to screen when a test was failing cox flake was interfering with it and printing everything but the reason the test failed. I'll have a look at GA see what flake we getting now, we should be in the clear anyway coz that issue was a while ago. Nice, Circle tests pass fine but why does Circle show the tests ran twice for your merge commit https://app.circleci.com/pipelines/github/ESMValGroup/ESMValCore?branch=main

**Comment by bouweandela** [comment]
The flake8 issue from the comment has been fixed, so I think we should be fine.

The tests ran twice because I clicked the button to run them again. I wanted to see if the caching helped to speed up the tests, but I found that I misunderstood how caching works on CircleCI. I made a new pull request to fix it: #1597. Could you have a look please?

**Comment by valeriupredoi** [comment]
ah that makes sense, I didn't have any beer today so was wondering why I see double :laughing: Yes, just approved that one!

**Pull request merged**